### PR TITLE
fix: TV media files card, and SQLite foreign key errors

### LINKED ISSUE #587: Follow-ups from #585: empty Media Files card on TV shows, and FK violations raising on SQLite
Two follow-ups identified during #585 that were deliberately left out of that PR's scope. Unrelated to each other; filed together to avoid two thin issues.

---

## 1. The Media Files card is permanently empty on every TV show page

`lib/mydia_web/live/media_live/show/components.ex:661` gates `media_files_section/1` on `length(@media_item.media_files) > 0` and iterates the same list at line 669.

A `MediaFile` belongs to **either** `media_item_id` (movies) **or** `episode_id` (TV episodes), never both, enforced by a check constraint. So `media_item.media_files` is always empty for a TV show, and episode files live at `media_item.episodes[].media_files`.

This is long-standing rather than new. What makes it worth fixing now is that #585 fixed the *adjacent* card: `subtitles_section/1` was switched to `Helpers.all_media_files/1` (`lib/mydia_web/live/media_live/show/helpers.ex:48`), which walks both sources. The result on a show page today is a working Subtitles card listing every episode's tracks, sitting directly beside a Media Files card that is always empty. The inconsistency is more visible than the original uniform limitation was.

**Fix:** use the same `Helpers.all_media_files/1` in `media_files_section/1`. It already exists, is already used by its neighbour, and its movie path is provably unchanged (`Loaders.build_preload_list/0` always preloads `episodes: [media_files: ...]`, so for a movie the episode list is empty and the helper reduces to exactly the prior list).

Worth checking whether anything else in that LiveView group has the same scoping. `components.ex:91` gates the playback block on `@media_item.type == "movie"` explicitly, so that one is intentional, but a sweep for `@media_item.media_files` would be sensible while in there.

Also worth a test with a TV fixture, since none of the existing tests for that section use one. That absence is why the subtitles version of this bug survived until review.

---

## 2. `ecto_sqlite3` swallows the foreign-key constraint name, so FK violations raise instead of returning a changeset error

Found while implementing `setSubtitleOffset` in #585 and fixed **only** at that one call site.

`ecto_sqlite3`'s `to_constraints/2` returns an unnamed `foreign_key` tuple. Ecto therefore cannot match it against a changeset's declared `foreign_key_constraint/2`, and re-raises `Ecto.ConstraintError` rather than returning `{:error, changeset}`. PostgreSQL is unaffected because it reports the real constraint name.

The user-visible effect: any resolver or context function that takes a client-supplied id and inserts a row referencing it returns a clean error on PostgreSQL and **raises a 500 on SQLite**, which is the default adapter for self-hosted installs.

Two id shapes hit it, and the second is the more common:
- A malformed id. On SQLite a `binary_id` is stored as TEXT, so `"garbage"` is a perfectly valid value and no cast error fires; the failure surfaces later as the FK violation.
- A **well-formed but nonexistent** id. This is the realistic one: a client holding a stale id after the referenced row was deleted.

The fix applied in #585, for reference, is in `Mydia.Subtitles.TrackSettings.set_offset/3` — a rescue scoped to `error.type == :foreign_key` with everything else re-raised, which is tight because that schema has exactly one foreign key.

During that work the mechanism was reported to affect roughly **17 similar call sites** across the codebase. **That count is unverified** and is the first thing worth establishing. The reviewer's recommendation, which I agree with, was that auditing and fixing them is its own piece of work rather than something to fold into a feature branch.

Suggested first step: grep for `foreign_key_constraint` declarations whose schemas are written from client-supplied ids, and check which of those paths are reachable from a GraphQL resolver, a REST controller, or a LiveView event handler. A shared helper may be a better answer than 17 individual rescues, since a per-site rescue is easy to forget on the eighteenth.

---

Neither blocks anything. The first is a visible UI inconsistency; the second is a latent 500 on the default adapter.

### PR DESCRIPTION
**Pull request by arsfeld**
Closes #587. Two unrelated follow-ups from #585, filed together in that issue and shipped together here.

## 1. The Media Files card was permanently empty on TV shows

A `MediaFile` belongs to either `media_item_id` (movies) or `episode_id` (TV episodes), never both, enforced by a check constraint. So `media_item.media_files` is always empty for a TV show, and `media_files_section/1` both gated and iterated on exactly that list. #585 fixed the neighbouring Subtitles card via `Helpers.all_media_files/1`, which left a working card sitting beside a permanently empty one.

The sweep the issue asked for found four more sites with the same root cause, three of them outside the file the issue named:

| Site | Symptom before |
|---|---|
| `show/components.ex` `media_files_section/1` | Card never rendered on a show page |
| `show/file_events.ex` `refresh_all_file_metadata/2` | "No media files to refresh" on every show page |
| `index.ex` `apply_quality_filter/2` | Any active quality filter silently excluded **every** TV show |
| `index.ex` `get_quality_badge/1` | Quality badge always blank on show cards |
| `index.ex` `total_file_size/1` | List view Size column always read "0 B" for shows |

All five now go through `Helpers.all_media_files/1`. The movie path is provably unchanged: `Loaders.build_preload_list/0` always preloads `episodes:`, so a movie's episode list is empty and the helper reduces to the prior list. The index page already preloaded `episodes: [media_files: ...]` on every path that feeds these functions, so this is in-memory work over data the page was loading anyway, with no new queries.

Worth noting for anyone repeating the sweep: grepping `media_item.media_files` misses all three `index.ex` sites, which bind the item to a local called `item`. Grep `.media_files`.

## 2. Foreign key violations raised a 500 on SQLite

`ecto_sqlite3` cannot recover which constraint failed from SQLite's error message, so its `to_constraints/2` maps every foreign key violation to `[foreign_key: nil]`. Ecto matches a changeset's declared constraints by name, a nil name matches nothing, and Ecto raises `Ecto.ConstraintError` instead of returning `{:error, changeset}`. PostgreSQL reports the real name and is unaffected.

The effect: any write taking a client-supplied id returned a clean error on PostgreSQL and raised a 500 on SQLite, which is the default adapter for self-hosted installs. The realistic trigger is not a malformed id but a well-formed one whose row was deleted, held by a stale client.

`Mydia.Repo` now wraps `insert`, `update` and `insert_or_update` with `Mydia.Repo.ForeignKeyGuard`. On a nil-named violation it walks the changeset's declared foreign keys, resolves each to its `belongs_to`, existence-checks the value, and rebuilds the error from the constraint entry so the caller sees the changeset PostgreSQL would have produced. Anything it cannot attribute is logged and reraised, so an undeclared constraint or a race stays as loud as it is today. No call site changed, and `Ecto.Multi` is covered because it dispatches through these functions.

**The guard is a strict no-op on PostgreSQL.** It handles only an error that is both `type: :foreign_key` and `constraint: nil`, and only `ecto_sqlite3` produces a nil name. The green PostgreSQL suite below is the evidence.

### A divergence left in place, deliberately

A *malformed* id, as opposed to a well-formed but nonexistent one, behaves differently and the test says so per adapter rather than papering over it. `Ecto.Type.cast_fun(:binary_id)` is a bare `is_binary` check on both adapters, so `"garbage"` produces a valid changeset either way. SQLite then stores it and fails the foreign key, which the guard converts. PostgreSQL instead fails in the `Ecto.UUID` dumper and raises `Ecto.ChangeError` before any constraint is evaluated. That is a type failure rather than a constraint failure, it predates this work, and the guard neither causes nor hides it.

### What the guard cannot cover

A write that takes no changeset has nothing to attach an error to. The audit went looking for those and found two live ones, both fixed here by checking ids up front:

- `Collections.add_items/2`, via `Repo.insert_all/3` with a client-supplied `media_item_id`.
- `Media.update_media_items_batch/2`, via `Repo.update_all/3` with a client-supplied `quality_profile_id` from the batch edit form. **This one raised on both adapters**, not just SQLite, so it is a plain 500 that predates the SQLite issue and would have outlived a fix aimed only at it. `Repo.update_all` has no changeset for Ecto to translate against, so it surfaced the raw `Exqlite.Error` rather than an `Ecto.ConstraintError`.

`plugins/kv.ex` already carried its own rescue and needed nothing.

### The audit

The issue reported "roughly 17" affected call sites and flagged the number as unverified. Establishing it was the first thing worth doing, and the answer reframes the work: **18 client-facing write paths carry a client-supplied foreign key, and 16 of them needed no edit at all**, because the guard covers them for free. That is the argument for fixing this at the repo level rather than per site. The two exceptions are the changeset-free writes above.

The per-site approach had already failed in practice, which is the clearest evidence available. #585 added the rescue to `set_offset/3`. #588 then added `record_resync/4` the following day with a verbatim copy, its comment noting it mirrors `set_offset/3`. Both are removed here.

## Testing

`test/mydia/repo/foreign_key_guard_test.exs` asserts the same result on both adapters for a nonexistent id, attribution across two foreign keys (a valid `collection_id` with a bogus `media_item_id` blames only the latter), an undeclared constraint still raising, an `Ecto.Multi` step halting rather than raising, and a valid insert being untouched, plus the malformed-id case which asserts each adapter's genuinely different outcome.

The equivalence check for removing #585 and #588's rescues is that `test/mydia/subtitles/track_settings_test.exs` passes **byte-for-byte unmodified** on both adapters: 20/20 before and after.

| | Result |
|---|---|
| SQLite full suite | 9,398 tests + 13 doctests, 0 failures |
| PostgreSQL full suite | 9,399 tests + 13 doctests, 0 failures |
| `mix credo --strict` | No issues, 1,407 files |
| `mix compile --warnings-as-errors` | Clean |

## Risk

`Ecto` marks only `default_options/1`, `prepare_query/3` and `prepare_transaction/2` as overridable. Overriding the write functions works because `use Ecto.Repo` defines them in the calling module, so `defoverridable` and `super` apply normally, but a major Ecto release could change the generated shape. The "a valid insert is untouched" test is first in the guard suite and fails immediately if that happens.

## Summary by CodeRabbit

* **New Features**
  * TV shows now include episode files in quality filters, badges, file sizes, metadata refreshes, and media-file displays.
  * Missing related records receive clearer validation feedback.

* **Bug Fixes**
  * Invalid collection items and media update references now return errors without partial changes.
  * Foreign-key failures are handled consistently instead of raising unexpected database errors.

* **Tests**
  * Added coverage for collection validation, media updates, foreign keys, and TV episode file handling.

### CODE REVIEW

File: lib/mydia_web/live/media_live/index.ex [hunk 1]
```diff
@@ -952,7 +958,10 @@ defmodule MydiaWeb.MediaLive.Index do
   defp format_year(year), do: year
 
   defp get_quality_badge(media_item) do
-    case media_item.media_files do
+    # A MediaFile belongs to either media_item_id (movies) or episode_id (TV
+    # episodes), never both, so media_item.media_files alone is always empty
+    # for a TV show and the badge was silently blank on every show card.
+    case MediaFileHelpers.all_media_files(media_item) do
```
[the remark this answers is not in the document]
Comment (arsfeld) [reply lib/mydia_web/live/media_live/index.ex:965]: Good catch, and confirmed. Elixir compares strings by codepoint, so `Enum.sort(["2160p", "720p"], :desc)` yields `["720p", "2160p"]` and the badge reported the worst episode as the show's quality.

Worth noting the bug predates this branch: a movie with several files of mixed resolution already hit it. What changed is the blast radius. Now that TV shows reach this function at all, mixed resolutions across a season make it ordinary rather than a corner case, so it belongs in this PR.

Fixed in b2d28cff3 using `DownloadService.parse_resolution_height/1` as suggested, plus a regression test with 2160p and 720p episodes asserting the badge shows 2160p and refutes 720p.

File: lib/mydia_web/live/media_live/index.ex [hunk 2]
```diff
@@ -952,17 +959,22 @@ defmodule MydiaWeb.MediaLive.Index do
...
+        # Rank by parsed height rather than lexically. Enum.sort(:desc) on the
+        # raw strings puts "720p" ahead of "2160p", and a show's episodes
+        # routinely mix resolutions, so that is easy to hit now that TV shows
+        # reach this function at all.
         files
         |> Enum.map(& &1.resolution)
         |> Enum.reject(&is_nil/1)
-        |> Enum.sort(:desc)
-        |> List.first()
+        |> Enum.max_by(&DownloadService.parse_resolution_height/1, fn -> nil end)
```
[the remark this answers is not in the document]
Comment (arsfeld) [reply lib/mydia_web/live/media_live/index.ex:977]: Checked this one directly rather than taking it on trust, and I'm going to decline it: the three-argument form is documented behavior, not an arity mistake.

`Enum.max_by/4` is `max_by(enumerable, fun, sorter \\ &>=/2, empty_fallback \\ ...)`, and it carries a clause that matches when the third argument is an arity-0 function, treating it as `empty_fallback` rather than as a sorter. So `Enum.max_by(list, fun, fn -> nil end)` resolves to exactly the call this suggestion proposes.

Verified on the pinned toolchain (Elixir 1.19.5, same version CI uses):

```
two elements: {:ok, "2160p"}
one element:  {:ok, "1080p"}
empty list:   {:ok, nil}
```

No `BadArityError`, correct maximum, and `nil` on the empty list, which is the behavior the previous `List.first/1` had. `mix compile --warnings-as-errors` is also clean, so there is no deprecation on this form.

Adding an explicit `&>=/2` would only restate the default, so I'd rather leave the shorter documented form. Happy to switch if you have a case where the three-argument clause does not apply.
